File: lua/bank-account/bank-account.lua
local BankAccount = {}

function BankAccount.new(self, o)
	self.funds = 0
	o = o or {}
	setmetatable(o, self)
	self.__index = self
	return o
end

function BankAccount.deposit(self, amount)
	assert(not self.closed)
	assert(amount > 0)
	self.funds = self.funds + amount
end

function BankAccount.balance(self)
	return self.funds
end

function BankAccount.withdraw(self, amount)
	assert(not self.closed)
	assert(amount > 0 and amount <= self.funds)
	self.funds = self.funds - amount
end

function BankAccount.close(self)
	self.funds = 0
	self.closed = true
end

return BankAccount

File: typescript/bank-account/bank-account.ts
export class ValueError extends Error {
  constructor() {
    super('Bank account error')
  }
}

export class BankAccount {
  private _balance = 0
  private active = false

  private checkIfActive(): void {
    if (!this.active) throw new ValueError()
  }

  open(): void {
    if (this.active) throw new ValueError()
    this.active = true
  }

  close(): void {
    this.checkIfActive()
    this._balance = 0
    this.active = false
  }

  deposit(amount: number): void {
    this.checkIfActive()
    if (amount < 0) throw new ValueError()
    this._balance += amount
  }

  withdraw(amount: number): void {
    this.checkIfActive()
    if (amount > this._balance || amount < 0) throw new ValueError()
    this._balance -= amount
  }

  get balance(): number {
    this.checkIfActive()
    return this._balance
  }

  set balance(_) {
    throw new Error('Bank account error')
  }
}
